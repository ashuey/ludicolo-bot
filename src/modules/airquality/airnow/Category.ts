export interface Category {
    Number: number
    Name: string
}
