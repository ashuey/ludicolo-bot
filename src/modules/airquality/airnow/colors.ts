import { HexColorString } from "discord.js";

export const AirQualityColors: Readonly<Record<number, HexColorString>> = Object.freeze({
    1: "#00e400",
    2: "#ffff00",
    3: "#ff7e00",
    4: "#ff0000",
    5: "#8f3f97",
    6: "#7e0023",
});
