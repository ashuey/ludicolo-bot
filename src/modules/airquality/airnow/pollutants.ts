export const betterPollutantNames: Readonly<Record<string, string>> = Object.freeze({
    'o3': 'Ozone',
    'pm2.5': 'Fine Particles'
});
