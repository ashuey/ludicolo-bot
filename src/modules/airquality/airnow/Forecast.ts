import { Category } from "@/modules/airquality/airnow/Category";

export interface Forecast {
    DateIssue: string;
    DateForecast: string;
    ReportingArea: string;
    StateCode: string;
    Latitude: number;
    Longitude: number;
    ParameterName: string;
    AQI: number;
    Category: Category;
    ActionDay: boolean;
    Discussion: string;
}
