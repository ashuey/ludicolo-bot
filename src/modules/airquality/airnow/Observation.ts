import { Category } from "@/modules/airquality/airnow/Category";

export interface Observation {
    DateObserved: string
    HourObserved: number
    LocalTimeZone: string
    ReportingArea: string
    StateCode: string
    Latitude: number
    Longitude: number
    ParameterName: string
    AQI: number
    Category: Category
}
