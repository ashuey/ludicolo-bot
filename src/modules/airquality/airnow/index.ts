import axios from "axios";
import { Observation } from "@/modules/airquality/airnow/Observation";
import { Result } from "@/common/Result";
import { Forecast } from "@/modules/airquality/airnow/Forecast";

export class AirNow {
    static readonly URL_BASE = "https://www.airnowapi.org";

    protected readonly http;

    constructor(apiKey: string) {
        this.http = axios.create({
            timeout: 2000,
            baseURL: AirNow.URL_BASE,
            params: {
                'API_KEY': apiKey,
                'format': 'application/json',
            }
        });
    }

    public async currentObservation(zip: string, distance = 50): Promise<Result<Observation[]>> {
        const result = await this.http.get<Observation[]>('aq/observation/zipCode/current', {
            params: {
                'zipCode': zip,
                'distance': distance,
            }
        });

        if (result.status < 200 || result.status >= 300) {
            return [false, new Error(`HTTP Error ${result.status}`)];
        }

        return [true, result.data];
    }

    public async historicalObservations(
        zip: string,
        date: string,
        distance = 50
    ): Promise<Result<Observation[]>> {
        const result = await this.http.get<Observation[]>('aq/observation/zipCode/historical', {
            params: {
                'zipCode': zip,
                'date': date,
                'distance': distance,
            }
        });

        if (result.status < 200 || result.status >= 300) {
            return [false, new Error(`HTTP Error ${result.status}`)];
        }

        return [true, result.data];
    }

    public async forecasts(
        zip: string,
        date: string | null = null,
        distance = 50
    ): Promise<Result<Forecast[]>> {
        const params: Record<string, string> = {
            'zipCode': zip,
            'distance': String(distance),
        };

        if (date) {
            params['date'] = date;
        }

        const result = await this.http.get<Forecast[]>('aq/forecast/zipCode', {
            params,
        });

        if (result.status < 200 || result.status >= 300) {
            return [false, new Error(`HTTP Error ${result.status}`)];
        }

        return [true, result.data];
    }
}
