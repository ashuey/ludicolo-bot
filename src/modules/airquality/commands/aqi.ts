import { bold, ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder } from "discord.js";
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import { AirQualityColors } from "@/modules/airquality/airnow/colors";
import { AirNowProvider } from "@/modules/airquality/AirNowProvider";
import { Command } from "@/common/Command";
import { Forecast } from "@/modules/airquality/airnow/Forecast";
import { Observation } from "@/modules/airquality/airnow/Observation";
import { betterPollutantNames } from "@/modules/airquality/airnow/pollutants";
import {logger} from "@/logger";

dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(customParseFormat);

export class AQICommand implements Command {
    // TODO: Don't hardcode this
    static readonly DEFAULT_ZIP = 14614;

    protected module: AirNowProvider

    constructor(module: AirNowProvider) {
        this.module = module;
    }

    build() {
        return (new SlashCommandBuilder())
            .setName('aqi')
            .setDescription('Gets the current Air Quality Index for a location')
            .addIntegerOption(option =>
                option.setName('zip_code')
                    .setDescription('5-digit ZIP Code')
                    .setMinValue(10000)
                    .setMaxValue(99999))
    }

    async execute(interaction: ChatInputCommandInteraction) {
        const zip = interaction.options.getInteger('zip_code') ?? AQICommand.DEFAULT_ZIP;

        const response = (new EmbedBuilder())
            .setFooter({ text: "Data provided by AirNow.gov" });

        const current = await this.getCurrentObservations(zip);
        let now = dayjs();

        if (current) {
            now = now.tz(current.LocalTimeZone);
        }

        const [today, tomorrow] = await Promise.all([
            this.getForecast(zip, now),
            this.getForecast(zip, now.add(1, 'day')),
        ])

        let title = "Air Quality";
        let aqi = 'Unavailable';
        let primaryPollutant = 'Unavailable';
        let todayStr = 'Unavailable';
        let tomorrowStr = 'Unavailable';
        let currentCategory = 'Unavailable';

        if (current) {
            const color = AirQualityColors[current.Category.Number];

            // Add location to the title
            title = title + ` in ${current.ReportingArea}`;
            currentCategory = `${color ? `${color[1]} ` : ''} ${current.Category.Name}`

            // Get the timestamp of the observation
            const timestamp = dayjs.tz(
                `${current.DateObserved.trim()} ${current.HourObserved}`,
                'YYYY-MM-DD HH',
                current.LocalTimeZone
            ).toDate();

            const pollutantKey = current.ParameterName.toLowerCase().trim();

            aqi = String(current.AQI);
            primaryPollutant = betterPollutantNames[pollutantKey] ?? current.ParameterName;

            response
                .setTimestamp(timestamp)
                .setColor(color ? color[0] : null);
        }

        if (today) {
            const emoji = AirQualityColors[today.Category.Number];
            todayStr = `${emoji ? emoji[1] : ''} ${today.Category.Name}`.trim();
        }

        if (tomorrow) {
            const emoji = AirQualityColors[tomorrow.Category.Number];
            tomorrowStr = `${emoji ? emoji[1] : ''} ${tomorrow.Category.Name}`.trim();
        }

        response
            .setTitle(title)
            .setDescription(currentCategory)
            .addFields(
                {
                    name: 'Air Quality Index',
                    value: aqi,
                    inline: true,
                },
                {
                    name: 'Primary Pollutant',
                    value: primaryPollutant,
                    inline: true,
                },
                {
                    name: 'Forecast',
                    value: `${bold('Today:')} ${todayStr}\n${bold('Tomorrow:')} ${tomorrowStr}`,
                },
            );

        return interaction.reply({embeds: [response]});
    }

    protected async getCurrentObservations(zip: number): Promise<Observation|null> {
        const [obResult, observations] = await this.module.airNow.currentObservation(String(zip));

        if (!obResult) {
            logger.warn(`Error getting air quality observations: ${observations}`);

            return null;
        }

        let observation: Observation | null = null;

        observations.forEach(o => {
            if (observation === null || o.AQI > observation.AQI) {
                observation = o;
            }
        })

        return observation;
    }

    protected async getForecast(zip: number, date: dayjs.Dayjs): Promise<Forecast|null> {
        const dateString = date.format('YYYY-MM-DD');
        const [fResult, forecasts] = await this.module.airNow.forecasts(String(zip), dateString);

        if (!fResult) {
            logger.warn(`Error getting air quality forecasts for ${dateString}: ${forecasts}`);

            return null;
        }

        const forecastsForDate = forecasts.filter(f =>
            f.DateForecast.toLowerCase().trim() === dateString
        );

        let forecast: Forecast | null = null;

        forecastsForDate.forEach(f => {
            if (forecast === null || f.AQI > forecast.AQI) {
                forecast = f;
            }
        })

        return forecast;
    }
}
