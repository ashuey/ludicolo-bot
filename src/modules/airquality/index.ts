import { Module } from "@/common/Module";
import { AQICommand } from "@/modules/airquality/commands/aqi";
import { Application } from "@/common/Application";
import { AirNow } from "@/modules/airquality/airnow";
import { AirNowProvider } from "@/modules/airquality/AirNowProvider";

export class AirQualityModule implements Module, AirNowProvider {
    readonly name = 'air_quality';

    readonly commands = [
        new AQICommand(this),
    ];
    readonly migrations = [];

    protected app: Application;

    protected _airNow: AirNow | undefined;

    constructor(app: Application) {
        this.app = app;
    }

    get airNow(): AirNow {
        if (!this._airNow) {
            this._airNow = new AirNow(this.app.config.airNowApiKey);
        }

        return this._airNow;
    }
}
