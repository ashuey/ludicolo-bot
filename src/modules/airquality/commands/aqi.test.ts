import {AQICommand} from "@/modules/airquality/commands/aqi";
import {AirNow} from "@/modules/airquality/airnow";

class AirNowTest extends AirNow {
    private callIdx = 0;

    private responses: unknown[] = [
        [{"DateObserved":"2024-07-26","HourObserved":13,"LocalTimeZone":"EST","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":28,"Category":{"Number":1,"Name":"Good"}},{"DateObserved":"2024-07-26","HourObserved":13,"LocalTimeZone":"EST","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":37,"Category":{"Number":1,"Name":"Good"}}],
        [{"DateIssue":"2024-07-26","DateForecast":"2024-07-26","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":32,"Category":{"Number":1,"Name":"Good"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-26","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":38,"Category":{"Number":1,"Name":"Good"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-27","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":48,"Category":{"Number":1,"Name":"Good"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-27","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":53,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-28","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":51,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-28","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":55,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-29","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":51,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-29","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":60,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""}],
        [{"DateIssue":"2024-07-26","DateForecast":"2024-07-27","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":48,"Category":{"Number":1,"Name":"Good"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-27","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":53,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-28","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":51,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-28","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":55,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-29","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"O3","AQI":51,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""},{"DateIssue":"2024-07-26","DateForecast":"2024-07-29","ReportingArea":"Eastern Lake Ontario Region","StateCode":"NY","Latitude":43.5263,"Longitude":-76.8398,"ParameterName":"PM2.5","AQI":60,"Category":{"Number":2,"Name":"Moderate"},"ActionDay":false,"Discussion":""}],
    ]

    setupTests() {
        const adapterFn = jest.fn(async (config) => {
            const responseData = this.responses[this.callIdx];
            this.callIdx += 1;
            return {
                data: responseData,
                status: 200,
                statusText: '',
                headers: {},
                config: config
            }
        });
        this.http.defaults.adapter = adapterFn;
        return adapterFn
    }
}

describe('aqi command', () => {
    afterEach(() => {
        jest.useRealTimers();
        jest.restoreAllMocks();
    })

    it('returns the expected response', async () => {
        jest.useFakeTimers();
        jest.setSystemTime(1722020988000) // Friday, July 26, 2024 3:09:48 PM GMT-04:00

        const replyFn = jest.fn();
        const airNow = new AirNowTest('');
        airNow.setupTests()
        const cmd = new AQICommand({
            airNow,
        });

        await cmd.execute({
            reply: replyFn,
            options: {
                getInteger: () => 14612,
            }
        } as never);

        expect(replyFn.mock.calls[0][0].embeds[0].toJSON()).toMatchSnapshot();
    })
})
