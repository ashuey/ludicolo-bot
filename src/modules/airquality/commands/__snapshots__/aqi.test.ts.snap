// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aqi command returns the expected response 1`] = `
{
  "color": 58368,
  "description": "ðŸŸ¢  Good",
  "fields": [
    {
      "inline": true,
      "name": "Air Quality Index",
      "value": "37",
    },
    {
      "inline": true,
      "name": "Primary Pollutant",
      "value": "Fine Particles",
    },
    {
      "name": "Forecast",
      "value": "**Today:** ðŸŸ¢ Good
**Tomorrow:** ðŸŸ¡ Moderate",
    },
  ],
  "footer": {
    "icon_url": undefined,
    "text": "Data provided by AirNow.gov",
  },
  "timestamp": "2024-07-26T18:00:00.000Z",
  "title": "Air Quality in Eastern Lake Ontario Region",
}
`;
